### PR TITLE
refactor: improve complexity of removing preselected coins

### PR DESCRIPTION
**Pull request by rag-hav**
Improve the complexity of removing pre-selected coins from the pool to linear from quadratic.

### CODE REVIEW

**Review by jonatack — COMMENTED** [review]
Concept ACK on replacing O(n) erase with O(1) operations push_back and swap, provided the difference is meaningful (a benchmark comparison might be helpful).

**Review by Eunoia1729 — COMMENTED** [review]
Concept ACK

**Review by mzumsande — COMMENTED** [review]
Could this be solved easier with the [erase-remove idiom](https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom) so you don't need to swap with a second vector?

**Comment by jonatack** [comment]
Or we could wait until C++20 to use constant-time `erase_if` https://en.cppreference.com/w/cpp/container/vector/erase2, as the currently available https://en.cppreference.com/w/cpp/algorithm/remove has complexity of std::distance(first, last) applications of the predicate.

**Comment by theStack** [comment]
Concept ACK

**Review by Mitumaru — APPROVED** [review]

**Comment by joshuadbryant1985** [comment]
Ok thank you

File: src/wallet/spend.cpp [hunk 1]
```diff
@@ -490,12 +490,14 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, const std::vec
...
-        else
-            ++it;
+    if(coin_control.HasSelected()){
+        std::vector<COutput> vCoins_filtered;
+        for (std::vector<COutput>::iterator it = vCoins.begin(); it != vCoins.end(); it++)
+        {
+            if (!preset_coins.count(it->outpoint))
+                vCoins_filtered.push_back(*it);
+        }
+        swap(vCoins_filtered, vCoins);
```
[the remark this answers is not in the document]
Comment (rag-hav) [reply src/wallet/spend.cpp]: made these changes

**Comment by rag-hav** [comment]
> Could this be solved easier with the [erase-remove idiom](https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom) so you don't need to swap with a second vector?

#24821

**Comment by rag-hav** [comment]
I am not sure what I am supposed to do right now. Do I wait for more reviewers?

**Comment by jonatack** [comment]
> I am not sure what I am supposed to do right now. Do I wait for more reviewers?

Haven't gotten back yet as there are many PRs to review; my suggestions would be the last entry of this section
https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#finding-reviewers and https://jonatack.github.io/articles/how-to-review-pull-requests-in-bitcoin-core.

**Comment by luke-jr** [comment]
>I am not sure what I am supposed to do right now.

If you wanted to use the remove-erase idiom, you should push that over this PR's branch.

**Comment by rag-hav** [comment]
> > I am not sure what I am supposed to do right now.
> 
> If you wanted to use the remove-erase idiom, you should push that over this PR's branch.

I am not sure if the remove-erase idiom is better, it saves memory but is harder to read as well. I prefer the current push_back and swap approach.

**Comment by mzumsande** [comment]
> I am not sure if the remove-erase idiom is better, it saves memory but is harder to read as well. I prefer the current push_back and swap approach.

Just FYI, I think it could be quite short like 
```
if(coin_control.HasSelected()){
    auto in_preset = [&preset_coins](COutput& x) { return preset_coins.count(x.outpoint); };
    vCoins.erase(std::remove_if(vCoins.begin(), vCoins.end(), in_preset), vCoins.end());
}
``` 
(didn't test this though) 
I guess whether that is easier to read depends on whether one is accustomed to the remove-erase idiom / lambdas or not.

**Comment by Eunoia1729** [comment]
> I am not sure if the remove-erase idiom is better, it saves memory but is harder to read as well. I prefer the current push_back and swap approach.  
   
As @jonatack suggested, we can also wait until C++20. It does both: saves memory and is easier to read.

**Comment by joshuadbryant1985** [comment]
> `-    if(coin_control.HasSelected()){
+    if (coin_control.HasSelected()) {
         std::vector<COutput> vCoins_filtered;
-        for (std::vector<COutput>::iterator it = vCoins.begin(); it != vCoins.end(); it++)
-        {
-            if (!preset_coins.count(it->outpoint))
+        for (std::vector<COutput>::iterator it = vCoins.begin(); it != vCoins.end(); ++it) {
+            if (!preset_coins.count(it->outpoint)) {
                 vCoins_filtered.push_back(*it);
+            }`

**Comment by achow101** [comment]
I would also prefer the erase-remove idiom as I think it's clearer as to what is actually happening (although perhaps having both erase and remove is a little bit confusing, but they're both clear that something is being deleted).

**Comment by mzumsande** [comment]
Looks like this can be closed, #24584 changed this spot to use the remove-erase idiom.

**Comment by furszy** [comment]
Worth to mention that #25685 completely removes the need to erase the pre-selected coins from the available coins vector by not including them there in the first place.

**Pull request closed without merging**